td-shim/src/bin/td-shim/e820.rs

```diff
@@ -158,27 +154,58 @@ impl E820Table {
     }
 
     fn merge(&mut self) {
-        let mut entry_num = self.entries.len();
         let mut idx = 0;
         loop {
-            if idx == entry_num - 1 {
+            if idx == self.num_entries - 1 {
                 break;
             }
             let entry_end = self.entries[idx].addr + self.entries[idx].size;
             if entry_end == self.entries[idx + 1].addr
                 && self.entries[idx].r#type == self.entries[idx + 1].r#type
             {
                 self.entries[idx].size += self.entries[idx + 1].size;
-                self.entries.remove(idx + 1);
-                entry_num -= 1;
+                self.remove_entry(idx + 1);
                 continue;
             }
             idx += 1;
         }
     }
 
+    fn insert_entry(&mut self, index: usize, entry: E820Entry) -> Result<(), E820Error> {
+        if self.num_entries >= MAX_E820_ENTRY {
+            return Err(E820Error::TooManyEntries);
+        }
+        if index > self.num_entries {
+            return Err(E820Error::InvalidIndex);
+        }
+
+        // Move all entries after the index back one position
+        for idx in (index..self.num_entries).rev() {
+            self.entries[idx + 1] = self.entries[idx];
+        }
+
+        self.entries[index] = entry;
+        self.num_entries += 1;
+        Ok(())
+    }
+
+    fn remove_entry(&mut self, index: usize) -> Result<(), E820Error> {
+        if index >= self.num_entries {
```

Comment:
What if self.num_entries is 0 ?

Reply:
if `self.num_entries == 0`, the `index >= self.num_entries` will always be true, then it will return error